[SYCL] Fix marray math function impls

JackAKirk opened the pull request:
This PR aims to fix issue : https://github.com/intel/llvm/issues/5991 and provide efficient working marray math function implementations for all backends.

marray math function support is currently switched on for {n} ({n} defined in #5991) but the implementations are currently broken and untested. There is also very limited test coverage for sycl::vec cases. The sycl 2020 specification states that the set {N} ({N} defined in #5991) should be supported for marray math function cases.

All SYCL 2020 math, native math, and half_precision math functions now have marray support when the function's arguments are of type `genfloat` and have the same argument type for all arguments.

Tests: https://github.com/intel/llvm-test-suite/pull/1002

Signed-off-by: jack.kirk <jack.kirk@codeplay.com>

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
I've added scalar_vector_* lists in this PR that omit marray types, so that math functions can distinguish the marray implementations I added.
The type lists including marrays, used in e.g. `is_genfloat`, are used in the `has_known_identity` trait class described in section 4.9.2. of the SYCL 2020 spec. The current marray lists include marrays of size from the set {n} (defined/discussed in https://github.com/intel/llvm/issues/5991) which limits the spans used in array reductions to the set {n}.  If we have array reductions then 4.9.2 does not state that they should be limited to the set {n}, although it does not specify what the admissible set of spans are.

I think that it makes more sense to allow array reductions with any span (or at least a larger range than {n}) which would mean updating the marray type lists.

@aobolensk @steffenlarsen what do you think?

JackAKirk commented:
> /verify with [intel/llvm-test-suite#1002](https://github.com/intel/llvm-test-suite/pull/1002)

FYI I don't have access to see the failures from this. The tests are passing locally for cuda.

steffenlarsen commented on sycl/include/CL/sycl/builtins.hpp:
```diff
@@ -27,95 +27,226 @@ namespace __sycl_std = __host_std;
+#define __SYCL_MATH_FUNCTION_2_OVERLOAD(NAME)                                  \
+  template <typename T, size_t N>                                              \
+  inline __SYCL_ALWAYS_INLINE std::enable_if_t<                                \
+      std::is_same<T, half>::value || std::is_same<T, float>::value ||         \
+          std::is_same<T, double>::value,                                      \
```
Could this be replaced with `detail::is_sgenfloat<T>`?

aelovikov-intel reviewed (commented):
Can we extend existing tests to capture the new sizes?

steffenlarsen commented:
> I've added scalar_vector_* lists in this PR that omit marray types, so that math functions can distinguish the marray implementations I added. The type lists including marrays, used in e.g. `is_genfloat`, are used in the `has_known_identity` trait class described in section 4.9.2. of the SYCL 2020 spec. The current marray lists include marrays of size from the set {n} (defined/discussed in #5991) which limits the spans used in array reductions to the set {n}. If we have array reductions then 4.9.2 does not state that they should be limited to the set {n}, although it does not specify what the admissible set of spans are.
> 
> I think that it makes more sense to allow array reductions with any span (or at least a larger range than {n}) which would mean updating the marray type lists.
> 
> @aobolensk @steffenlarsen what do you think?

I agree, `genfloat` is currently too restrictive on `marray` and we should loosen it. If we did, would this patch be obsolete or would these separate definitions still be required?

aelovikov-intel commented on sycl/include/CL/sycl/builtins.hpp:
```diff
@@ -27,95 +27,220 @@ namespace __sycl_std = __host_std;
+
+#undef __SYCL_MATH_FUNCTION_OVERLOAD
+
+#define __SYCL_MATH_FUNCTION_2_OVERLOAD(NAME)                                  \
```
Would something like this https://godbolt.org/z/adez76fTd be possible here to avoid duplicating this code for all 3 cases?

JackAKirk replied:
Yes I think that something along that line would be possible, although I'm not sure that it would necessarily be an improvement, particularly for the current implementation.

aelovikov-intel replied:
Having seen (almost?) the same `for` loop nine times in this PR, I'd argue it will be. Other reviewers, am I really the only one?

JackAKirk replied:
There's a couple of issues with applying the suggested approach in the simple example https://godbolt.org/z/adez76fTd. Firstly the two `__invoke_*` function calls that are made in each `marray` math function implementation require the explicit provision of a template parameter that is different in each of the calls, i.e. `vec<T, 2>` and `T`. This means that an immediate adaption of https://godbolt.org/z/adez76fTd that directly called the `__invoke_*` functions would required the provision of two lambdas, `Callable Fvec` and `Callable F` corresponding to the `vec<T, 2>` and `T` cases.
Also with this approach the usage of macros to prevent duplicating function declaration lines becomes less attractive because this would require passing the lambdas to the macros, although I imagine that you probably meant to remove the macros completely.

I brought this issue up with the team and everyone agreed that although it would be possible to work around these issues the resultant implementation would be more complex and the size of the code would be similar to the current implementation.

I have applied your change to `to_vec2` that simplifies it.

Thanks

JackAKirk commented on sycl/include/CL/sycl/builtins.hpp:
```diff
@@ -27,95 +27,220 @@ namespace __sycl_std = __host_std;
+      std::enable_if_t<detail::is_sgenfloat<T>::value, sycl::marray<T, N>>     \
+      NAME(sycl::marray<T, N> x) __NOEXC {                                     \
+    sycl::marray<T, N> res;                                                    \
+    auto x_vec2 = reinterpret_cast<sycl::vec<T, 2> const *>(&x);               \
```
Yes this usage of reinterpret_cast is UB in C++. I've now switched to using `std::memcpy` instead which leads to identical asm at the default Opt level.

JackAKirk commented on sycl/include/CL/sycl/builtins.hpp:
```diff
@@ -27,95 +27,220 @@ namespace __sycl_std = __host_std;
+    sycl::marray<T, N> res;                                                    \
+    auto x_vec2 = reinterpret_cast<sycl::vec<T, 2> const *>(&x);               \
+    auto res_vec2 = reinterpret_cast<sycl::vec<T, 2> *>(&res);                 \
+    for (size_t i = 0; i < N / 2; i++) {                                       \
```
This implementation was chosen for simplicity and because it is the most optimal implementation for the CUDA backend for the functions which can use specialized f16x2 instructions that are defined here: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#half-precision-floating-point-instructions. 

It would be possible to choose some different implementations that made use of vectors of size 16, 8, and 4 for example, but, without having a good understanding of how different implementations would affect different backends, I decided that this simple implementation was best, at least as an initial solution. I'm not sure how the logarithmic loop would work in this case, since the largest `sycl::vec` size is 16.

JackAKirk replied:
I've added a comment to this effect in the code: note that using larger vector sizes could also be better for the CUDA backend: it wouldn't prevent the libclc implementation of the vector cases from using the vector of size two ptx instructions for the operations because this is what happens already for the vector implementations. 
However as I noted here: https://github.com/intel/llvm/pull/6038#discussion_r891673257 the load/store operations are not being vectorized after compilations in the case that we construct the vectors from an marray. I planned to investigate why the compiler is doing this after adding an initial implementation (since I imagined a not perfectly optimized working implementation is better than no implementation), but I can look into why the compiler is not vectorizing the loads/stores now.

JackAKirk commented on sycl/include/sycl/ext/oneapi/experimental/builtins.hpp:
```diff
@@ -115,6 +142,31 @@ inline __SYCL_ALWAYS_INLINE
+  for (size_t i = 0; i < N / 2; i++) {
+    auto partial_res = __sycl_std::__invoke_exp2<sycl::vec<half, 2>>(
+        sycl::detail::to_vec(x, i * 2));
+    std::memcpy(&res[i * 2], &partial_res, sizeof(vec<half, 2>));
```
Do you mean an explicit conversion operator allowing something like 
```
marray<T, N> x;
marray<vec<T, 2>, N/2> x_vec2 = x;
``` 
or similar.

Or an implicit conversion operator from `x[i * 2]` to `vec<T, 2>` etc?

I think that the main issue (beyond the question of whether this would make the code more readable or less confusing) is that this would have to break SYCL spec definitions of  marray/vec/etc? Perhaps I am missing something or misunderstanding what you mean?

Thanks

dkhaldi replied:
marray is supposed to be more general than vec. 
What is the long term plan with vec? Will marray replace vec?
@Pennycook, can you comment on that?

JackAKirk replied:
So this marray implementation supports any marray size N where N is in the set of size_t: so it is more general than vec in this sense. I think that it makes sense to map the marray implementation onto the vec one like e.g. I have done here by using vec2 (although it would be a fair argument to suggest another impl that makes use of larger vec sizes, as we've discussed in another thread). In theory this should also allow for vectorized register loads, although for some reason when we cast from marray to vec the vectorized loads that we see when using the standard vec implementation are not used: This is something we have as a TODO to improve on.

keryell replied:
There was a good presentation by your colleagues at syclcon.org this year that talks about it: *Untangling Modern Parallel Programming Models* https://www.youtube.com/watch?v=6FbW6zVYkxk&list=PL46sP9LM8GsyHAxj1k7MbWrv5f5SlMpIF&index=27

bader replied:
> marray is supposed to be more general than vec. What is the long term plan with vec? Will marray replace vec? @Pennycook, can you comment on that?

+@rolandschulz

rolandschulz replied:
We added `marray` in SYCL 2020 to untangle `sycl::vec`. The problem with `sycl::vec` is that it serves two different use case: explicit SIMD programming (vectorization across it with subgroup size of 1) and a simple math array (vectorization across the subgroup). And those two use cases require different API (e.g. you don't want element wise access for a SIMD vector that's why `sycl::vec` has swizzle). For the SIMD use-case, DPC++ has for now ESIMD. For SYCL 2023 we would like to add a new class aligned with `std::simd` and deprecate `sycl::vec`.

JackAKirk replied:
> Do you mean an explicit conversion operator allowing something like
> 
> ```
> marray<T, N> x;
> marray<vec<T, 2>, N/2> x_vec2 = x;
> ```
> 
> or similar.
> 
> Or an implicit conversion operator allowing `x[i * 2]` -> `vec<T, 2>` etc?
> 
> I think that the main issue (beyond the question of whether this would make the code more readable or less confusing) is that this would have to break SYCL spec definitions of marray/vec/etc? Perhaps I am missing something or misunderstanding what you mean?
> 
> Thanks

@keryell do you have any thoughts on the above?

Thanks

keryell replied:
> Do you mean an explicit conversion operator allowing something like
> 
> ```
> marray<T, N> x;
> marray<vec<T, 2>, N/2> x_vec2 = x;
> ```
> 
> or similar.
> 
> Or an implicit conversion operator allowing `x[i * 2]` -> `vec<T, 2>` etc?
> 
> I think that the main issue (beyond the question of whether this would make the code more readable or less confusing) is that this would have to break SYCL spec definitions of marray/vec/etc? Perhaps I am missing something or misunderstanding what you mean?

Yes, this would be the idea.
I was just thinking here to make your extension code clearer, by having an implicit linearization of the recursive types.
Whether we expose this officially in SYCL is another story. But if enough people think is useful, why not?

JackAKirk replied:
I didn't think it was really possible to have such conversions without exposing it in SYCL since I had thought C++ requires such conversion operators to be defined within the class, and if they could be defined outside of the class then it could be hard for someone to find the definition of the conversion operator. Conversions between `vec<T, 2>` and `marray<T, N>` or `T` look strange to me and I wouldn't know what it is doing without looking up the conversion operator.

jchlanda replied:
> I think that it makes sense to map the marray implementation onto the vec one like e.g. I have done here by using vec2 (although it would be a fair argument to suggest another impl that makes use of larger vec sizes, as we've discussed in another thread). In theory this should also allow for vectorized register loads, although for some reason in the CUDA backend when we cast from marray to vec the vectorized loads that we see when using the standard vec implementation are not used: This is something we have as a TODO to improve on.

We've discussed this a bit offline.
The reason why loads and stores to/from `marray`s are not being vectorized is that LLVM's `load-store-vectorizer` pass has a strict requirement on the alignment; the pointers have to be aligned to at least what the resulting vector would require (or the target must allow misaligned operations). The alignment requirement can be easily achieved by changing the default alignment of `marray` to the "previous" vector (i.e.: `marray<15, T>` would be decorated with `__attribute__((aligned(8 * sizeof(T)))`, making sure it's a power of 2 number).

WRT behind the scenes conversion of `marray` elements to vectors (`to_vec2`), it seems wrong, it is unlikely to bring performance benefits, as it will always result in temporary storage and extra loads, stores instructions. Perhaps we could use the same approach as cutlas does, harnessing the fact that alignment is set correctly, we could do a bit of type punning see: https://github.com/NVIDIA/cutlass/blob/e7a61c761a4bfb387b61c03cdbcd19ab300726b7/include/cutlass/functional.h#L1444

I had a go at the above in here: https://github.com/JackAKirk/llvm/pull/1

As a side note, it feels like somewhere here there is a compiler optimization missed, we should be able to use the same logic as the vectorizer pass and gather those scalar intrinsic, converting them to vector equivalents, making this code a lot cleaner.

JackAKirk replied:
@aelovikov-intel @keryell 

What do you think about @jchlanda 's suggestion above?

aelovikov-intel replied:
I don't think it becomes less UB.

> it seems wrong, it is unlikely to bring performance benefits, as it will always result in temporary storage and extra loads, stores instructions.

Why wouldn't the compiler optimize this? This is a standard C++ idiom for the compiler - https://en.cppreference.com/w/cpp/numeric/bit_cast

jchlanda replied:
> Why wouldn't the compiler optimize this? This is a standard C++ idiom for the compiler - https://en.cppreference.com/w/cpp/numeric/bit_cast

The compiler mustn't optimize it away because of the difference in the alignment of (elements of) `marray` and the temporary `vec2` variable, as `marray` follows the `std::array` alignment rules. For the same reason compiler was unable to generate vector loads and stores directly to/from `marray`, this is enforced in PTX spec:

> By default, vector variables are aligned to a multiple of their overall size (vector length times base-type size), to enable vector load and store instructions which require addresses aligned to a multiple of the access size.

This patch changes it though, and `marray` satisfies the alignment requirement of the "previous" vector. 

> I don't think it becomes less UB.

You are right, I reverted the type punning casts, with alignment fix LLVM is clever enough to optimize `vec2` variable and a call to `memcpy` away.

JackAKirk replied:
> 
> > I don't think it becomes less UB.
> 
> You are right, I reverted the type punning casts, with alignment fix LLVM is clever enough to optimize `vec2` variable and a call to `memcpy` away.

Sounds like we have a winner.

steffenlarsen commented on sycl/include/CL/sycl/builtins.hpp:
```diff
@@ -1395,175 +1530,284 @@ select(T a, T b, T2 c) __NOEXC {
+      __NOEXC {                                                                \
+    marray<float, N> res;                                                      \
+    for (size_t i = 0; i < N / 2; i++) {                                       \
+      auto partial_res = __sycl_std::__invoke_half_##NAME<vec<float, 2>>(      \
```
Since we know these are floats, is there a performance benefit to do the vector packing?

JackAKirk replied:
Yes I think so: in theory we should be able to use vectorized loads/stores.

JackAKirk replied:
I investigated this a bit further:

Using the OpenCL CPU backends vectorized loads are used in the final llvm IR when copying from marray to vector (including for float). It may be that the limitation of the loads/stores not being vectorized is only happening for NVPTX. I have therefore created an internal issue to try to fix this.

Some more information on the issue:

E.g. vector of size two implementation of tanh leads to this ptx for the fp16 case:

```
ld.global.b32 %hh1, [%rd4];
tanh.approx.f16x2 %hh2, %hh1;
mov.b32 {%h1, %h2}, %hh2;
st.global.v2.b16 [%rd4], {%h1, %h2};
```

Note this is little cleaner when using float since vec 2 loads for f32 are used rather than b32 for f16x2: then it will also use ld.global.v2.f32.

Whereas e.g. an marray of size two implementation of tanh (https://github.com/intel/llvm/pull/6038/files#diff-525a5c115806d68d90c778133cb3f137d979196033de778a085f30414e2bdf78R114) leads to (it should ideally match the above vector case):

 
```
ld.global.u16 %rd9, [%rd4];
ld.global.u16 %rd10, [%rd4+2];
shl.b64 %rd11, %rd10, 16;
or.b64 %rd12, %rd11, %rd9;
cvt.u32.u64 %r1, %rd12;
mov.b32 %hh1, %r1;
tanh.approx.f16x2 %hh3, %hh1;
mov.b32 {%h1, %h2}, %hh3;
st.global.b16 [%rd4], %h1;
st.global.b16 [%rd4+2], %h2;

```

From examining the llvm IR this appears to be because the compiler loses information about the copy from marray to vector during the optimization, then it loses the vectorized loads and stores. I think this is happening for all marray math functions.
I'm guessing that this is not something that is trivial to fix, and probably doesn't really make much difference. So I don't think this particular issue should block this PR.

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/builtins.hpp:
```diff
@@ -100,10 +100,37 @@ inline __SYCL_ALWAYS_INLINE
+  sycl::marray<T, N> res;
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
+  for (size_t i = 0; i < N / 2; i++) {
+    auto partial_res = __clc_native_tanh(sycl::detail::to_vec2(x, i * 2));
```
`native::tanh` above does some conversion to OCL types before calling `__clc_native_tanh`. Should this also be done here? Could this just call `native::tanh` and rely on the vector implementation of that?

JackAKirk replied:
So I tested this with the CUDA and OpenCL backends and found no difference whether we use these conversion functions or not: from what I read there isn't a reason to cast from float to cl_float if you are only dealing with device code although I may have misunderstood and there may be some devices where this is necessary; for consistency I've used the OCL conversion stuff via calling the native::tanh vector implementations as you suggest.

steffenlarsen commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

pvchupin commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

pvchupin commented:
[2022-06-28T02:21:02.113Z] Failed Tests (2):
[2022-06-28T02:21:02.113Z]   SYCL :: DeviceLib/half_precision_math_test_marray_vec.cpp
[2022-06-28T02:21:02.113Z]   SYCL :: DeviceLib/native_math_test_marray_vec.cpp

JackAKirk commented:
> [2022-06-28T02:21:02.113Z] Failed Tests (2): [2022-06-28T02:21:02.113Z] SYCL :: DeviceLib/half_precision_math_test_marray_vec.cpp [2022-06-28T02:21:02.113Z] SYCL :: DeviceLib/native_math_test_marray_vec.cpp

Interesting. I can't reproduce these failures on my local machine but I can reproduce the half_precision_math_test_marray_vec.cpp on the A100 and the behaviour depends on the cudatoolkit version used. 
I have not been able to reproduce a failure from native_math_test_marray_vec.cpp. It would be useful to see the errors from the CI failure but I don't have access to http://llvm-ci2.intel.com:8080/job/SYCL_CI/job/intel/job/LLVM_Test_Suite_Associate_CI/214/display/redirect. Would it be possible to provide the redirected link that I think I should have access to? Thanks.

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
> > [2022-06-28T02:21:02.113Z] Failed Tests (2): [2022-06-28T02:21:02.113Z] SYCL :: DeviceLib/half_precision_math_test_marray_vec.cpp [2022-06-28T02:21:02.113Z] SYCL :: DeviceLib/native_math_test_marray_vec.cpp
> 
> Interesting. I can't reproduce these failures on my local machine but I can reproduce the half_precision_math_test_marray_vec.cpp on the A100 and the behaviour depends on the cudatoolkit version used: the test gets incorrect values for a single case "half_precision_math_test_2_powr". I have not been able to reproduce a failure from native_math_test_marray_vec.cpp. It would be useful to see the errors from the CI failure but I don't have access to http://llvm-ci2.intel.com:8080/job/SYCL_CI/job/intel/job/LLVM_Test_Suite_Associate_CI/214/display/redirect. Would it be possible to provide the redirected link that I think I should have access to? Thanks.

@pvchupin I identified the problem with "half_precision_math_test_2_powr" to be an existing bug (that we will investigate) that was there before this PR and uncovered by the increased test coverage for float3 introduced in https://github.com/intel/llvm-test-suite/pull/1002. I've removed this test coverage but the CI is still failing and I can't reproduce the failure. Could you please post the redirected link from http://llvm-ci-test2.intel.com:8080/job/SYCL_CI/job/intel/job/LLVM_Test_Suite_Associate_CI/184/display/redirect so that I can see the failure? I don't have access to this link ending "redirect".  Thanks

pvchupin commented:
@JackAKirk, we use to have public dumps for these I think. @tfzhu can you confirm/check please if these are shutdown today?
We need to move this stuff to Github Actions...

Here is the output:
half_precision_math_test_marray_vec.cpp:
Fails on both OpenCL and Level Zero
```
[2022-06-28T02:24:58.807Z] # command stderr:
[2022-06-28T02:24:58.807Z] half_precision_math_test_marray_vec.cpp.tmp.out: /run/jenkins-dir/workspace/SYCL_CI/intel/Lin/LLVM_Test_Suite/llvm-test-suite/SYCL/DeviceLib/half_precision_math_test_marray_vec.cpp:57: void half_precision_math_test_exp10(sycl::queue &, T, T, size_t) [T = sycl::vec<float, 3>]: Assertion `checkEqual(result, ref)' failed.
```

native_math_test_marray_vec.cpp:
This one fails on OpenCL only
```
[2022-06-28T02:11:51.727Z] # command stderr:
[2022-06-28T02:11:51.728Z] native_math_test_marray_vec.cpp.tmp.out: /run/jenkins-dir/workspace/SYCL_CI/intel/Lin/LLVM_Test_Suite/llvm-test-suite/SYCL/DeviceLib/native_math_test_marray_vec.cpp:55: void native_math_test_exp2(sycl::queue &, T, T, size_t) [T = sycl::vec<float, 3>]: Assertion `checkEqual(result, ref)' failed.
[2022-06-28T02:11:51.728Z] 
[2022-06-28T02:11:51.728Z] error: command failed with exit status: -6
```

We have also SYCL :: USM/depends_on.cpp failing on Windows only, but not sure if it's related. Nothing useful though in the output...

```
[2022-06-30T11:14:48.285Z] $ "env" "SYCL_DEVICE_FILTER=opencl:cpu,host" "W:\jenkins-dir\workspace\SYCL_CI\intel\Win\Test_Suite\build\SYCL\USM\Output\depends_on.cpp.tmp1.out"
[2022-06-30T11:14:48.285Z] note: command had no output on stdout or stderr
[2022-06-30T11:14:48.285Z] error: command failed with exit status: 3221225477
```

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
> @JackAKirk, we use to have public dumps for these I think. @tfzhu can you confirm/check please if these are shutdown today? We need to move this stuff to Github Actions...
> 
> Here is the output: half_precision_math_test_marray_vec.cpp: Fails on both OpenCL and Level Zero
> 
> ```
> [2022-06-28T02:24:58.807Z] # command stderr:
> [2022-06-28T02:24:58.807Z] half_precision_math_test_marray_vec.cpp.tmp.out: /run/jenkins-dir/workspace/SYCL_CI/intel/Lin/LLVM_Test_Suite/llvm-test-suite/SYCL/DeviceLib/half_precision_math_test_marray_vec.cpp:57: void half_precision_math_test_exp10(sycl::queue &, T, T, size_t) [T = sycl::vec<float, 3>]: Assertion `checkEqual(result, ref)' failed.
> ```
> 
> native_math_test_marray_vec.cpp: This one fails on OpenCL only
> 
> ```
> [2022-06-28T02:11:51.727Z] # command stderr:
> [2022-06-28T02:11:51.728Z] native_math_test_marray_vec.cpp.tmp.out: /run/jenkins-dir/workspace/SYCL_CI/intel/Lin/LLVM_Test_Suite/llvm-test-suite/SYCL/DeviceLib/native_math_test_marray_vec.cpp:55: void native_math_test_exp2(sycl::queue &, T, T, size_t) [T = sycl::vec<float, 3>]: Assertion `checkEqual(result, ref)' failed.
> [2022-06-28T02:11:51.728Z] 
> [2022-06-28T02:11:51.728Z] error: command failed with exit status: -6
> ```
> 
> We have also SYCL :: USM/depends_on.cpp failing on Windows only, but not sure if it's related. Nothing useful though in the output...
> 
> ```
> [2022-06-30T11:14:48.285Z] $ "env" "SYCL_DEVICE_FILTER=opencl:cpu,host" "W:\jenkins-dir\workspace\SYCL_CI\intel\Win\Test_Suite\build\SYCL\USM\Output\depends_on.cpp.tmp1.out"
> [2022-06-30T11:14:48.285Z] note: command had no output on stdout or stderr
> [2022-06-30T11:14:48.285Z] error: command failed with exit status: 3221225477
> ```

It seems that there are some problems with some half_precision and native math functions with generic implementations when using float3 (these are all existing problems in sycl tip that the increased test coverage is revealing), although I can't reproduce the particular failures that you listed above locally. I've removed native_math_test_exp2 and half_precision_math_test_exp10 float3 test cases from https://github.com/intel/llvm-test-suite/pull/1002 but Jenkins/llvm-test-suite is still failing (I still can't see the results).  Is it now only failing for USM/depends_on.cpp? USM/depends_on.cpp should have no dependence on PR changes here.

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
I have now also added the -ffast-math flag support for the new marray cases, in line with the scalar/vector cases : https://github.com/intel/llvm/pull/5801 . This also affects/motivates the macro usage in builtins.hpp. The corresponding fast-math-flag.cpp test has been updated in https://github.com/intel/llvm-test-suite/pull/1002 to cover the new cases.
I've also added a device-code test here via native-math-cuda.cpp. This test checks that the correct native math instructions are called for the cuda backend. The precision of these native math functions is detailed in table 10 here: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#intrinsic-functions.
I thought that it would be sensible to simply check that these functions are being correctly called rather than making fresh tests that check the quoted native precision.
I have removed the more basic and general native (and half_precision) math function tests (a la https://github.com/intel/llvm-test-suite/blob/intel/SYCL/DeviceLib/math_test.cpp) that I originally added in https://github.com/intel/llvm-test-suite/pull/1002 because they are inappropriate for the opencl backend because the precision of the opencl backend is not precise enough to pass the tests. I think that at some point there should be added some more rigorous tests such as exist in the opencl CTS (https://github.com/KhronosGroup/OpenCL-CTS/blob/2d93b122c3078cd67a0528ad9e791dbcadaf03d6/test_common/harness/errorHelpers.cpp).

Anyway, the marray cases added here have passing test coverage at least matching the existing scalar/vector cases, plus the CUDA backend now has the native math builtins device code check.

JackAKirk commented:
I've just managed to access the full llvm-test-suite results and seen the windows failure due to the change to marray.hpp:324. Looking into this now...

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

steffenlarsen commented on sycl/include/sycl/ext/oneapi/experimental/builtins.hpp:
```diff
@@ -108,10 +108,41 @@ inline __SYCL_ALWAYS_INLINE
+tanh(sycl::marray<T, N> x) __NOEXC {
+  sycl::marray<T, N> res;
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
+#define FUNC_VEC native::tanh
```
Could you please add a comment on why NVPTX needs to use native here and for exp2?

JackAKirk replied:
Sure I can add an appropriate comment in the code

For reviewers reference: `sycl::tanh` and `sycl::exp2` are already supported for nvptx in the sycl/builtins.hpp header. The `sycl::native::tanh` does not exist in the SYCL spec and so it is added here: `sycl::ext::oneapi::experimental::tanh`. `sycl::native::exp2` does exist in the SYCL spec, however not for the half type, so the half case for exp2 is added here also.
The reason why NVPTX wants to use native implementations tanh and exp2 is, i guess, because they are frequently used functions.
The reason why we want to add them as an extension is 
1) The above reason
and
2) We want to support all ptx 7.0 instructions.

JackAKirk replied:
I'll also add a note saying that only nvptx backend supports `native::tanh`/`native::ex2`(for half) currently and other backends are set to fall back on `sycl::tanh` here.

JackAKirk commented:
> I've just managed to access the full llvm-test-suite results and seen the windows failure due to the change to marray.hpp:324. Looking into this now...

I've partially dealt with this but there are still failures on windows that I need to fix.

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
Hi @steffenlarsen 
So I've fixed the windows alignment issue and added comments for why the native::tanh and native::exp2(half) are calling non-native sycl::tanh and sycl::exp2 for backends other than nvptx.

The sycl marray/vec math test is failing for level_zero only on windows. I've tried to investigate it locally but I can't reproduce the failure, and the CI output isn't very helpful. So I've now marked the test unsupported for windows && level_zero: https://github.com/intel/llvm-test-suite/pull/1002/commits/cfc1e91c3c93ed3330fa1543de8ab2cac8eb55de

We could merge it as is in order to add marray math functions for the next release. If we also want level_zero && windows working straight away then it might take a bit more time.

Cheers!

steffenlarsen commented:
@romanovvlad - Do you know if changing the alignment of `marray` could affect ABI?

JackAKirk commented:
Just realized that it also failed on (opencl && windows) for the same test - I'll also mark that case UNSUPPORTED.

romanovvlad commented:
> @romanovvlad - Do you know if changing the alignment of `marray` could affect ABI?

I think so, let's consider the following:

```
some.hpp
  void foo(std::array<sycl::marray_WITHOUT_PATCH<char[3], 1>, 4> &);

old.cpp // compiled without the patch
  void foo(std::array<sycl::marray_WITHOUT_PATCH<char[3], 1>, 4> &);   

new.cpp // compiled with the patch

  void bar() {
     std::array<sycl::marray_WITHOUT_PATCH<char[3], 1>, 4> ArrayWithoutPatch;
     std::array<sycl::marray_WITH_PATCH<char[3], 1>, 4> ArrayWithPatch;
     assert(sizeof(ArrayWithoutPatch) == 4 * 3); // before the patch
     assert(sizeof(ArrayWithPatch)  == 4 * 4);
     foo(ArrayWithoutPatch);
     foo(ArrayWithPatch); // Callee is not aware of padings in the array.
  }
```

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
> > @romanovvlad - Do you know if changing the alignment of `marray` could affect ABI?
> 
> I think so, let's consider the following:
> 
> ```
> some.hpp
>   void foo(std::array<sycl::marray_WITHOUT_PATCH<char[3], 1>, 4> &);
> 
> old.cpp // compiled without the patch
>   void foo(std::array<sycl::marray_WITHOUT_PATCH<char[3], 1>, 4> &);   
> 
> new.cpp // compiled with the patch
> 
>   void bar() {
>      std::array<sycl::marray_WITHOUT_PATCH<char[3], 1>, 4> ArrayWithoutPatch;
>      std::array<sycl::marray_WITH_PATCH<char[3], 1>, 4> ArrayWithPatch;
>      assert(sizeof(ArrayWithoutPatch) == 4 * 3); // before the patch
>      assert(sizeof(ArrayWithPatch)  == 4 * 4);
>      foo(ArrayWithoutPatch);
>      foo(ArrayWithPatch); // Callee is not aware of padings in the array.
>   }
> ```

Do you think I should change anything in the implementation? Or does this mean it can only be merged at a certain time?

Thanks

steffenlarsen commented:
> Do you think I should change anything in the implementation? Or does this mean it can only be merged at a certain time?

I worry it is the latter.

pvchupin commented:
We can start using label "abi-break" I just created to mark and gather all changes waiting for the next window

steffenlarsen commented:
@JackAKirk - Is there some subset of these changes we can include without the alignment changes?

JackAKirk commented:
> @JackAKirk - Is there some subset of these changes we can include without the alignment changes?

Good point - I think that the whole thing can work without the alignment, it will just not vectorize loads and stores so won't be as fast. I can remove the alignment changes and run the tests to check there are no issues.

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

steffenlarsen commented:
> Good point - I think that the whole thing can work without the alignment change, it will just not vectorize loads and stores so won't be as fast. I can remove the alignment changes and run the tests to check there are no issues.

I see you have removed the alignment changes. Would you mind creating a PR with the reverted changes? They may rot a little before we get to the next ABI-break window, but if nothing it will act as a reminder to do it.

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

steffenlarsen reviewed (approved):
LGTM! Would still like to have the comment from https://github.com/intel/llvm/pull/6038/files#r973197193.

JackAKirk commented:
> LGTM! Would still like to have the comment from https://github.com/intel/llvm/pull/6038/files#r973197193.

Oh, I thought I added it

> LGTM! Would still like to have the comment from https://github.com/intel/llvm/pull/6038/files#r973197193.

I did add comments on the scalar functions: https://github.com/intel/llvm/blob/4824d63e89d15348acd9a93ceeee0f7df7b91433/sycl/include/sycl/ext/oneapi/experimental/builtins.hpp#L144 and https://github.com/intel/llvm/blob/4824d63e89d15348acd9a93ceeee0f7df7b91433/sycl/include/sycl/ext/oneapi/experimental/builtins.hpp#L97

I can also the same comment here https://github.com/intel/llvm/blob/4824d63e89d15348acd9a93ceeee0f7df7b91433/sycl/include/sycl/ext/oneapi/experimental/builtins.hpp#L118 ?

Or perhaps I also misunderstood the comment you meant to add? 

Thanks

steffenlarsen commented:
I think the comments you mentioned are good, but there seem to be more functions using native functions for NVPTX but not for other targets, like the place where the aforementioned comment is. May not be clear that there is a link between the other definitions using native for NVPTX only and the new ones doing the same.

JackAKirk commented:
> I think the comments you mentioned are good, but there seem to be more functions using native functions for NVPTX but not for other targets, like the place where the aforementioned comment is. May not be clear that there is a link between the other definitions using native for NVPTX only and the new ones doing the same.

I see what you mean. I've added the comment in the two other places now.

JackAKirk commented:
Hi @keryell 
Do you have any more reviews for this?
Thanks

JackAKirk commented:
> Hi @keryell Do you have any more reviews for this? Thanks

Also we want this in before we finish up the marray complex extension: https://github.com/intel/llvm/pull/6550.

JackAKirk commented:
> > Hi @keryell Do you have any more reviews for this? Thanks
> 
> Also we want this in before we finish up the marray complex extension: #6550.

@bader Do you want to ask someone else to review this perhaps?

bader commented:
> @bader Do you want to ask someone else to review this perhaps?

No. I requested @keryell to review to make sure that his previous comments are addressed.
If you think that you received all needed approves, I suggest we merge it and address further comments in follow-up commits.

keryell reviewed (approved):
Thanks.
I recently noticed https://github.com/KhronosGroup/SYCL-Docs/issues/320 by the way, so the `marray` are not yet a drop-in replacement for `vec`.

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1002

JackAKirk commented:
> > @bader Do you want to ask someone else to review this perhaps?
> 
> No. I requested @keryell to review to make sure that his previous comments are addressed. If you think that you received all needed approves, I suggest we merge it and address further comments in follow-up commits.

OK. I'm now happy for this PR and https://github.com/intel/llvm-test-suite/pull/1002 to be merged now. I've merged the latest sycl branch here, resolved conflicts and fixed one test that was resultantly failing. It is now passing the llvm-test-suite run using https://github.com/intel/llvm-test-suite/pull/1002. The amd test failures are I think unrelated and also seen in other PRs.

Thanks.
